### PR TITLE
User management

### PR DESCRIPTION
**Pull request by AvihayMReblaze**

### CODE REVIEW

File: src/router/index.ts [hunk 1]
```diff
@@ -291,6 +292,14 @@ const routes: RouteRecordRaw[] = [
               title: 'Publish Changes',
             },
           },
+          {
+            path: 'account',
+            name: 'Account',
+            component: AccountDetails,
+            meta: {
+              title: 'Account',
```
Comment (Aviv-Galmidi) [inline src/router/index.ts:300]:
```suggestion
              title: 'Account Details',
```

File: src/stores/userStore.ts [hunk 2]
```diff
@@ -228,6 +228,12 @@ export const useUserStore = defineStore('user', () => {
     return cookieRemainingTime <= 10 * 60 // less than 10 minutes remain on cookie expiration
   }
 
+  function getCookieAccountId():string {
```
Comment (Aviv-Galmidi) [inline src/stores/userStore.ts:231]:
```suggestion
  function getCookieAccountId(): string {
```

File: src/types/index.d.ts [hunk 3]
```diff
@@ -457,6 +457,31 @@ declare module CuriefenseClient {
 
   // Operation documents - END
 
+  // User Details - START
+
+  type Key = {
```
Comment (Aviv-Galmidi) [inline src/types/index.d.ts:462]: Key is a very very generic term, we should use a much more speicifc name for this type
```suggestion
  type UserAPIKey = {
```

File: src/types/index.d.ts [hunk 4]
```diff
@@ -457,6 +457,31 @@ declare module CuriefenseClient {
...
+    account_id: string
+    created_by: string
+    created_time: int
```
Comment (Aviv-Galmidi) [inline src/types/index.d.ts:465]:
```suggestion
    created_time: number
```

File: src/views/AccountDetails.vue [hunk 5]
```diff
@@ -0,0 +1,409 @@
+<template>
+    <div class="card-content">
+      <div class="media">
+        <div class="media-content">
+          <div class="columns">
+            <div class="column">
+            </div>
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:7]: Why do we need this empty column?

File: src/views/AccountDetails.vue [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+            <div class="column">
+              <div class="field is-grouped is-pulled-right">
+                <p class="control">
+                  <button class="button is-small save-document-button"
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:11]:
```suggestion
                  <button class="button is-small save-account-details-button"
```

File: src/views/AccountDetails.vue [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+                  </button>
+                </p>
+              </div>
+            </div>
+          </div>
+        </div>
+      </div>
+      <hr/>
+      <div v-show="!loadingDocCounter && selectedBranch && accountId">
+        <div class="content collapsible-content px-5 py-5">
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:28]: This div isn't collapsible

```suggestion
        <div class="content">
```

File: src/views/AccountDetails.vue [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+          <div class="columns columns-divided">
+            <div class="column is-4">
+              <div class="field">
+                <label class="label is-small">
+                  Your Name
+                </label>
+                <div class="control suffix">
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:35]: The input doesn't have a suffix

```suggestion
                <div class="control">
```

File: src/views/AccountDetails.vue [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+                         title="Your name"
+                         placeholder="Name"
+                         :disabled="parseInt(userDetails.acl) > 20">
+                  </div>
+              </div>
+              <div class="field">
+                <label class="label is-small">
+                  Your Email
+                </label>
+                <div class="control suffix">
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:48]: The input doesn't have a suffix

```suggestion
                <div class="control">
```

File: src/views/AccountDetails.vue [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+                         title="User Email"
+                         placeholder="Email"
+                         disabled>
+                </div>
+              </div>
+              <div class="field">
+                <label class="label is-small">
+                  Phone Number
+                </label>
+                <div class="control suffix">
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:61]: The input doesn't have a suffix

```suggestion
                <div class="control">
```

File: src/views/AccountDetails.vue [hunk 11]
```diff
@@ -0,0 +1,409 @@
...
+                         title="Phone Number"
+                         placeholder="Phone Number"
+                         :disabled="parseInt(userDetails.acl) > 20">
+                </div>
+              </div>
+              <div class="field">
+                <label class="label is-small">
+                  Organization Name
+                </label>
+                <div class="control suffix">
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:74]: The input doesn't have a suffix

```suggestion
                <div class="control">
```

File: src/views/AccountDetails.vue [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+                  </span>
+                </div>
+                <div class="content collapsible-content px-5 py-5">
+                  <div class="field is-flex center-qr">
+                    <div class="control">
+                      <canvas ref="refqrcode"></canvas>
+                    </div>
+                  </div>
+                  <p class="has-text-centered mt-4">
+                    <code>{{userDetails?.otpseed?.replace(/(.{4})(.{4})(.{4})(.{4})/, '$1 $2 $3 $4')}}</code>
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:114]: Let's move this functional code to a computed value
"userDetails?.otpseed?.replace(/(.{4})(.{4})(.{4})(.{4})/, '$1 $2 $3 $4')"

File: src/views/AccountDetails.vue [hunk 13]
```diff
@@ -0,0 +1,409 @@
...
+          </button>
+        </div>
+      </div>
+      <div v-if="generateKeyShown"
+           class="modal is-active">
+        <div class="modal-background">
+          <div class="modal-card is-size-7">
+            <header class="modal-card-head">
+              <h5 class="modal-card-title is-size-6 mb-0">
+                Delete user
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:184]: v-if says `generateKeyShown` but the text here is `Delete user`

File: src/views/AccountDetails.vue [hunk 14]
```diff
@@ -0,0 +1,409 @@
...
+          classes: 'width-160px',
+          cellContentClasses: 'ellipsis',
+          isSortable: true,
+          isSearchable: true,
+        },
+      ] as ColumnOptions[],
+      rowButtonIcon: 'fa-trash',
+      rowButtonTitle: 'Revoke key',
+      loadingDocCounter: 0,
+      showQr: false,
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:286]: Not being used
```suggestion
```

File: src/views/AccountDetails.vue [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+      isQrCollapsed: true,
+      isApiKeyCollapsed: true,
+      userDetails: {} as UserDetails,
+      apiKeys: [],
+      interval: null,
+      keysInterval: null,
+      accountId: null,
+      generateKeyShown: false,
+      isLoading: false,
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:295]: Should this be merged into `loadingDocCounter`?

File: src/views/AccountDetails.vue [hunk 16]
```diff
@@ -0,0 +1,409 @@
...
+      }
+    },
+
+    loadAccountId() {
+      this.interval = setInterval(() => {
+        this.loadAccountIdAfterInterval()
+      }, 500)
+      this.loadAccountIdAfterInterval()
+    },
+
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:331]: Can be removed

```suggestion
```

File: src/views/AccountDetails.vue [hunk 17]
```diff
@@ -0,0 +1,409 @@
...
+      const method = 'GET'
+      const response = await RequestsUtils.sendReblazeRequest({
+        methodName: method,
+        url,
+      })
+      this.userDetails = response?.data || {}
+      this.setLoadingDocStatus(false)
+    },
+
+    async loadKeys() {
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:344]: loadAPIKeys

File: src/views/AccountDetails.vue [hunk 18]
```diff
@@ -0,0 +1,409 @@
...
+        url,
+        data: {title: this.keyTitle},
+      })
+      this.generateKeyShown = false
+      this.keyTitle = null
+      this.loadKeys()
+      this.setLoadingDocStatus(false)
+    },
+
+    async revokeKey(id:string) {
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:376]: removeAPIKey

File: src/views/AccountDetails.vue [hunk 19]
```diff
@@ -0,0 +1,409 @@
...
+      otpauth,
+      {
+        width: 300,
+      },
+    )
+  },
+})
+</script>
+<style scoped lang="scss">
+  .center-qr {
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:406]: We can use `is-justify-content-center` class instead

**Review by Aviv-Galmidi — CHANGES_REQUESTED** [review]

File: src/views/AccountDetails.vue (hunk 19 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:406]: Done

File: src/views/AccountDetails.vue (hunk 18 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:376]: Done

File: src/views/AccountDetails.vue [hunk 20]
```diff
@@ -0,0 +1,409 @@
...
+      })
+      this.apiKeys = response?.data || []
+      this.setLoadingDocStatus(false)
+    },
+
+    openGenerateKeyModal() {
+      this.generateKeyShown = true
+    },
+
+    async requestToGenerateKey() {
```
[the remark this answers is not in the document]
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:360]: Done

File: src/views/AccountDetails.vue (hunk 17 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:344]: Done

File: src/views/AccountDetails.vue [hunk 21]
```diff
@@ -0,0 +1,409 @@
...
+  methods: {
+    setLoadingDocStatus(isLoading: boolean) {
+      if (isLoading) {
+        this.loadingDocCounter++
+      } else {
+        this.loadingDocCounter--
+      }
+    },
+
+    loadAccountIdAfterInterval() {
```
[the remark this answers is not in the document]
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:316]: Done

File: src/views/AccountDetails.vue (hunk 16 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:331]: Done

File: src/router/index.ts (hunk 1 above)
Comment (AvihayMReblaze) [reply src/router/index.ts:309]: Done

File: src/stores/userStore.ts (hunk 2 above)
Comment (AvihayMReblaze) [reply src/stores/userStore.ts:237]: Done

File: src/types/index.d.ts (hunk 3 above)
Comment (AvihayMReblaze) [reply src/types/index.d.ts:491]: Done

File: src/types/index.d.ts (hunk 4 above)
Comment (AvihayMReblaze) [reply src/types/index.d.ts:494]: Done

File: src/views/AccountDetails.vue (hunk 5 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:7]: We didn't need, removed and done.

File: src/views/AccountDetails.vue (hunk 6 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:11]: Done

File: src/views/AccountDetails.vue (hunk 8 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:35]: Removed, done.

File: src/views/AccountDetails.vue (hunk 9 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:48]: Removed, done.

File: src/views/AccountDetails.vue (hunk 11 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:74]: Removed, done.

File: src/views/AccountDetails.vue (hunk 13 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue]: Fixed, done.

File: src/views/AccountDetails.vue (hunk 14 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:286]: Removed, done.

File: src/views/AccountDetails.vue [hunk 22]
```diff
@@ -0,0 +1,409 @@
...
+      rowButtonIcon: 'fa-trash',
+      rowButtonTitle: 'Revoke key',
+      loadingDocCounter: 0,
+      showQr: false,
+      isQrCollapsed: true,
+      isApiKeyCollapsed: true,
+      userDetails: {} as UserDetails,
+      apiKeys: [],
+      interval: null,
+      keysInterval: null,
```
[the remark this answers is not in the document]
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:292]: Done.

File: src/views/AccountDetails.vue [hunk 23]
```diff
@@ -0,0 +1,409 @@
...
+      accountId: null,
+      generateKeyShown: false,
+      isLoading: false,
+      keyTitle: '',
```
[the remark this answers is not in the document]
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:296]: Done

File: src/views/AccountDetails.vue [hunk 24]
```diff
@@ -0,0 +1,409 @@
...
+    }
+  },
+  computed: {
+    selectedBranch(): string {
+      return this.branchesStore.selectedBranchId
+    },
+
+    ...mapStores(useBranchesStore),
+    ...mapStores(useUserStore),
```
[the remark this answers is not in the document]
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:305]: Done

File: src/views/AccountDetails.vue (hunk 12 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:114]: Done

File: src/views/AccountDetails.vue (hunk 15 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:295]: Changed to isGenerateLoading

File: src/views/AccountDetails.vue (hunk 10 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:61]: Done

File: src/views/AccountDetails.vue [hunk 25]
```diff
@@ -181,7 +179,7 @@
           <div class="modal-card is-size-7">
             <header class="modal-card-head">
               <h5 class="modal-card-title is-size-6 mb-0">
-                Delete user
+                Generate key
```
Comment (Aviv-Galmidi) [inline src/views/AccountDetails.vue:182]:
```suggestion
                Generate API key
```

**Review by Aviv-Galmidi — CHANGES_REQUESTED** [review]
Pleae fix unit tests as well

File: src/views/AccountDetails.vue (hunk 25 above)
Comment (AvihayMReblaze) [reply src/views/AccountDetails.vue:182]: Done

**Review by Aviv-Galmidi — APPROVED** [review]

**Pull request merged**